mismi-s3/src/Mismi/S3/Commands.hs

```diff
@@ -169,19 +169,83 @@ read' a = do
 
 copy :: Address -> Address -> AWS ()
 copy =
-  copyWithMode Fail
+  copyWithMode Overwrite
 
 copyWithMode :: WriteMode -> Address -> Address -> AWS ()
 copyWithMode mode s d = do
+  -- FIX error
   unlessM (exists s) . throwM $ SourceMissing CopyError s
+  -- FIX error
   foldWriteMode  (whenM (exists d) . throwM . DestinationAlreadyExists $ d) (pure ()) mode
-  copy' s d
+  size' <- getSize s
+  -- FIX error
+  size <- fromMaybeM (fail $ "Can not caluclate size of source object [" <> (T.unpack $ addressToText s) <> "].") size'
+  let chunk = 100 * 1024 * 1024 -- 100 mb
+      big = 1024 * 1024 * 1024 -- 1 gb
+  case size < big of
+    True ->
+      copy' s d
+    False ->
+      copyMultipart s d size chunk 100
 
 copy' :: Address -> Address -> AWS ()
 copy' (Address (Bucket sb) (Key sk)) (Address (Bucket b) (Key dk)) =
   void . send $ copyObject (BucketName b) (sb <> "/" <> sk) (ObjectKey dk)
      & coServerSideEncryption .~ Just sse & coMetadataDirective .~ Just Copy
 
+copyMultipart :: Address -> Address -> Int -> Int -> Int -> AWS ()
+copyMultipart source dest size chunk fork = do
+  mpu <- createMultipartUpload dest -- target
+  e <- ask
+
+  let chunks = calculateChunksCapped size chunk 4096
+
+  r <- liftIO $
+    consume (forM_ chunks . writeQueue) fork $ multipartCopyWorker e mpu source dest
+
+  let abort =
+        abortMultipart' dest mpu
+
+  case r of
+    Left f ->
+      abort >>
+        fail ("multipart copy failed in worker: " <> show f)
+
+    Right prts -> do
+      -- Sort is required here because the completeMultipartUpload api expects an
+      -- ascending list of part id's
+      let z = sortOn (\(PartResponse i _) -> i) $ snd prts
```

Comment:
Worth sharing this whole block with `multipartUpload`, they look identical to me?